adapters/hono-adapter-amqp-vertx/src/main/java/org/eclipse/hono/adapter/amqp/VertxBasedAmqpProtocolAdapter.java

```diff
@@ -0,0 +1,382 @@
+            }
+            return Future.<Void> succeededFuture();
+        }).recover(t -> {
+            LOG.info("Cannot process message for [deviceId: {}] of [tenantId: {}]", context.getDeviceId(),
+                    context.getTenantId());
+            if (AmqpErrorException.class.isInstance(t)) {
+                AmqpErrorException error = (AmqpErrorException) t;
+                context.reject(error);
+            } else if (!context.delivery().remotelySettled()) {
+                // client wants to be informed that the message cannot be processed.
+                ProtonHelper.released(context.delivery(), true);
+            }
+            return Future.failedFuture(t);
+        });
+    }
+
+    /**
+     * Closes the specified receiver link.
+     * 
+     * @param receiver The link to close.
+     */
+    private void onLinkDetach(final ProtonReceiver receiver) {
+        LOG.debug("closing link [{}]", receiver.getName());
+        receiver.close();
+    }
+
+    /**
+     * This method validates that a client tries to publish a message to a supported endpoint. If the endpoint is
+     * supported, the method returns a succeeded future containing the valid address.
+     * 
+     * @param receiver The link with which the message is received.
+     * @return A future with the address upon success or a failed future.
+     */
+    Future<ResourceIdentifier> validateEndpoint(final ProtonReceiver receiver) {
+        Future<ResourceIdentifier> result = Future.future();
+        ResourceIdentifier address = ResourceIdentifier.fromString(receiver.getRemoteTarget().getAddress());
+        switch (EndpointType.fromString(address.getEndpoint())) {
+        case TELEMETRY:
+            if (!ProtonQoS.AT_LEAST_ONCE
+                    .equals(receiver.getRemoteQoS()) && !ProtonQoS.AT_MOST_ONCE.equals(receiver.getRemoteQoS())) {
+                result.fail(new AmqpErrorException("unsupported-delivery-mode",
+                        "Telemetry endpoint only supports AT_LEAST_ONCE and AT_MOST_ONCE delivery semantics."));
+            } else {
+                // This adapter only supports AT_MOST_ONCE for telemetry?
+                receiver.setQoS(ProtonQoS.AT_MOST_ONCE);
+                result.complete(address);
+            }
+            break;
+        case EVENT:
+            if (!ProtonQoS.AT_LEAST_ONCE.equals(receiver.getRemoteQoS())) {
+                result.fail(new AmqpErrorException("unsupported-delivery-mode",
+                        "Event endpoint only supports AT_LEAST_ONCE delivery semantics."));
+            } else {
+                receiver.setQoS(ProtonQoS.AT_LEAST_ONCE);
+                result.complete(address);
+            }
+            break;
+        default:
+            LOG.error(String.format("Endpoint with [name: %s] is not supported by this adapter ",
+                    address.getEndpoint()));
+            result.fail(new AmqpErrorException("unsupported-delivery-mode",
```

Comment:
unsupported-delivery-mode ? This should probably result in the standard `amqp:not-found` error, shouldn't it?